src/vm/types/serialization.rs

```diff
@@ -0,0 +1,862 @@
+    format!(
+        r#"{{ "type": "{}", "value": {} }}"#,
+        type_name, val)
+}
+
+impl ClaritySerializable for Value {
+    fn serialize(&self) -> String {
+        use super::Value::*;
+        use super::PrincipalData::*;
+
+        match self {
+            Int(value) => json_simple_object(TYPE_I128, &to_hex(value)),
+            Buffer(value) => json_simple_object(TYPE_BUFF, &hash::bytes_to_hex(&value.data)),
+            Bool(value) => {
+                let str_value = if *value {
+                    "true"
+                } else {
+                    "false"
+                };
+                json_recursive_object(TYPE_BOOL, str_value)
+            },
+            Principal(StandardPrincipal(data)) => {
+                json_simple_object(TYPE_STANDARD_PRINCIPAL, &data.to_address())
+            },
+            Principal(ContractPrincipal(simple_name)) => {
+                // NOTE: this should eventually panic, since "unresolved" contract principals
+                //       should never be materialized to the database.
+                format!(
+                    r#"{{ "type": "{}", "issuer": ":none:", "name": "{}" }}"#,
+                    TYPE_CONTRACT_PRINCIPAL, simple_name)
+            },
+            Principal(QualifiedContractPrincipal{ sender, name }) => {
+                format!(
+                    r#"{{ "type": "{}", "issuer": "{}", "name": "{}" }}"#,
+                    TYPE_CONTRACT_PRINCIPAL, sender.to_address(), name)
+            },
+            Response(response) => {
+                let type_name = if response.committed {
+                    TYPE_RESP_OK
+                } else {
+                    TYPE_RESP_ERR
+                };
+                let value = response.data.serialize();
+                json_recursive_object(type_name, &value)
+            },
+            Optional(OptionalData{ data: None }) => {
+                format!(r#"{{ "type": "{}" }}"#, TYPE_OPT_NONE)
+            },
+            Optional(OptionalData{ data: Some(value) }) => {
+                json_recursive_object(TYPE_OPT_SOME, &value.serialize()) 
+           },
+            List(data) => {
+                let entries: Vec<String> = data.data
+                    .iter().map(|x| x.serialize())
+                    .collect();
+                let entries_str = entries.join(", ");
+                format!(
+                    r#"{{ "type": "{}", "entries": [ {} ] }}"#,
+                    TYPE_LIST, entries_str)
+            },
+            Tuple(data) => {
+                let entries: Vec<String> = data.data_map
+                    .iter().map(|(key, value)|
+                                format!(r#""{}": {}"#, key, value.serialize()))
+                    .collect();
+                let entries_str = entries.join(", ");
+                format!(
+                    r#"{{ "type": "{}", "entries": {{ {} }} }}"#,
+                    TYPE_TUPLE, entries_str)
+            }
+        }
+    }
+}
```

Comment:
I think this is the case, but I just want to be sure it's been considered:  there's _no possible way_ for a value to serialize to anything that has JSON special characters, right?  i.e. no `,`, `{`, `[`, `}`, `]`, or `"`?

Reply:
That's correct -- the only "strings" that are serialized are tuple names, and currently, the parser will not allow those to contain any of those values. However, like the other Strings in use through the VM, we should have guarded String types for these values.

Reply:
Sounds good.  As long as #1078 will introduce a guarded string type, then I think we're good here.